### DISCUSSION TITLE
Will spec add HEARTBEAT message?

### DISCUSSION THREAD

**Discussion by SunFulong** [wamp-proto/wamp-proto#490] [Q&A] [answered]
As I know, browser's websocket client can't send websocket native ping/pong message, so if a websocket have a short idle timeout (10 mins), connection will drop frequently, will spec add HEARTBEAT message to keep alive the connection?

**Comment by konsultaner** [comment]
@SunFulong WebSockets have `HEARTBEAT`s and so do RawSockets. Just look for `ping` and `pong`

**Reply by SunFulong** [reply]
> @SunFulong WebSockets have `HEARTBEAT`s and so do RawSockets. Just look for `ping` and `pong`

I know ping/pong, but web browser can't send out `ping` message

**Reply by KSDaemon** [reply, marked as answer]
You're right. There is no way around that. 
For now you have a few options:

1. Setup ping messages on the router side as some implementations allow that.
2. Setup fake RPC in your backend and call it regularly from the browser client side or even just publish to some meaningless topic with confirmation (thus you can be sure that both directions work: r/w)

IDK about introducing ping/pong into the WAMP protocol itself... As WAMP — is an Application level protocol. And it is not caring much about transport problems (to some degree). And btw, the problem is only with one specific transport and only in one type of environment: browser.

**Reply by oberstet** [reply]
Yes, to the best of my knowledge:

* there is no W3C API defined to control WebSocket ping/pong control frames from JavaScript or other W3C API users
* there is no way to configure Google Chrome / Mozilla Firefox so that it automatically sends WebSocket ping control frames on WebSocket connections opened from JavaScript
* all browsers support _responding_ to received WebSocket pings though

That means, heartbeating on WebSocket needs to be done manually, at the application level, using regular WebSocket data frames.

For realiable heartbeating, those need to be initiated from two sides .. from the router and from the client.

As @KSDaemon notes, some WAMP routers allow automatic router initiated WebSocket ping control frames.

For the other direction (client initiated): you need app level, as described above.

E.g. do acknowledged WAMP publications, or perform some WAMP RPC regularily ..

Since you can use any acknowledged publication (the "pong" is the WAMP ACK for the publication), I don't see much pressure to standardize that in WAMP currently ...

**Reply by SunFulong** [reply]
> Setup ping messages on the router side as some implementations allow that.

@KSDaemon , I will going to setup PING and PONG message with message code 256 and 257, implement it like ICMP.

**Reply by SunFulong** [reply]
> E.g. do acknowledged WAMP publications, or perform some WAMP RPC regularily

@oberstet , Why not make a real PING/PONG message, exchange some runtime datas, such as each side timestamp.

**Reply by KSDaemon** [reply]
First of all, it will be backward incompatible %) So any brand new client sending a PING message will end with its connection being aborted.

**Reply by KSDaemon** [reply]
We are at that stage when there are plenty of implementations and we can not so easily introduce incompatible changes. That's why we always take care of adding new options which in most cases can be ignored if not supported.

**Reply by konsultaner** [reply]
I don't see a need for `PING` `PONG` on WAMP level. My router just sends those messages and I never had any issues with that. If you really want to check on the client side you can also use RPC to get a `PING` `PONG`. Adding extra message types would blow up implementations for no reason. Maybe add an option to the `CALL` but please no more message types.

**Reply by SunFulong** [reply]
> That's why we always take care of adding new options which in most cases can be ignored if not supported.

@KSDaemon , This time, I really understand you. If we are moving next version `wamp.3` , this may be native support, but `wamp.2` , we only can use existing model to implement new requirements.

**Reply by SunFulong** [reply]
> I don't see a need for `PING` `PONG` on WAMP level.

@konsultaner , you missed some of the previous discussion. My background is implement a router running on Amazon Serverless Application Model, their WebSocket API Gateway have too many limitations, I'm trying to resolve such limitations, one is their transport have idle timeout (10 mins) and hard connection time limit (2 hours). So, I need `PING` or other `HEARTBEAT` method to avoid idle timeout.

**Reply by konsultaner** [reply]
@SunFulong I see what you mean and totally understand, but your usecase is quite special. I still think you should work with existing messages to emulate this behaviour. @om26er idea is actually quite a good one.

**Reply by SunFulong** [reply]
@konsultaner Yes, I'm going to using om26er's solution.

However, I don't think my use case is special, cloud compute it is a solution. WAMP is a great protocol and I've developed several commercial systems with it. Now, I'm trying to host them in the cloud, so there are a series of compatibility issues to discuss.

**Reply by oberstet** [reply]
mmh, what do you mean with "compatibility issues"? WAMP is fully compatible with e.g. WebSocket as defined in RFC6455 as a transport. so there is nothing to fix in WAMP (in this regard) and no need for a `wamp.3`?

IMO, keeping a WAMP transport alive is a transport concern, and so the first natural layer to look at is the transport.

as @konsultaner mentions, both WebSocket and RawSocket have ping/pong built in. eg RawSocket

> For a PING message (TTT == 1), the length is the length of the arbitrary payload that follows. A Peer MUST reply to each PING by sending exactly one PONG immediately, and the PONG MUST echo back the payload of the PING exactly.

https://wamp-proto.org/wamp_ap_latest_ietf.html#section-7.1-61

----

having said that, the solution @om26er suggested is often what is done in practice ... but the topic you use to publish with acknowledgement as well as the event payload published is not really important ... but the spec might of course discuss and suggest something to choose. also discuss the timing and timeouts that should be used.

> there are networks where the _amount_ (bytes) of keep alive messages is important - when no other (non keep alive, real) traffic is present, the network might shift down the connection to a low-energy/high-latency profile .. eg mobile networks are doing that. so you want at least one message every N seconds _and_ at least M bytes/s traffic (== heartbeat payload) bandwidth on average

**Reply by SunFulong** [reply]
> what do you mean with "compatibility issues"?

I mean, there is a compatibility issue running on AWS, not WAMP itself.

There is no problem with the WAMP specification itself, but it has strict requirements for the runtime environment (such as a WebSocket server that fully complies with RFC). This cannot be said to be a mistake, but as a developer, the runtime environment chosen may not fully meet the requirements of the specification (such as AWS). Therefore, I have sent several discussions about how to make WAMP compatible with AWS.

> so there is nothing to fix in WAMP (in this regard) and no need for a wamp.3

Of course, nothing needs to be fixed, but, such as PING, PONG, HEARTBEAT or KEEP-ALIVE and any other way to reduce requirement stringency, I think they can be added to `wamp.3` .

> both WebSocket and RawSocket have ping/pong built in

Yes, but web browsers do not support it. We can't say: don't use WAMP in web browsers, then we need to find a way to be compatible with them.

**Comment by oberstet** [comment]
https://github.com/wamp-proto/wamp-proto/blob/master/rfc/text/basic/bp_03_messages.md#extension-messages allows

> Implementations MAY define and use implementation specific messages with message type codes from the extension message range [256, 1023]. 

what the current spec text might lack to mention explicitly for extension messages:

* how this is announced by a peer
* how a peer not supporting the announced use of extension messages should do
* what a peer not understanding extension messages used should do (even though it did not deny to connect having seen the announcement)

**Comment by om26er** [comment]
One workaround that I use to "emulate" a ping on top of WAMP is to publish to a topic like `foo.bar.echo` with `acknowledge` set to true. That essentially means the communication between the router and the client is intact and functional.

**Reply by SunFulong** [reply]
I'm a little shaky, maybe using a dummy topic is a better approach.
